Use libnftables in dynamically linked binary

robmry opened the pull request:
**- What I did**

When dockerd is dynamically linked, use cgo to call functions in libnftables instead of exec-ing the nft binary.

On my M2 macbook, using the library knocks about 20ms off each update, down to less than 1ms. Initialisation and teardown operations happen sequentially, so the time saving comes directly off the total.

Using libnftables brings nftables update performance in line with iptables - typical times ...

|  | iptables | nft | libnftables |
|-|---------|----|-----------|
|docker network create --ipv6 b46 | 48ms | 57ms | 10ms |
|docker run --rm -ti --network b46 -p 8080:80 busybox (start) | 95ms | 185ms | 109ms |

<details>
<summary>Click for OTEL traces ...</summary>

### Network create

#### iptables

<img width="1487" height="561" alt="network-create-iptables" src="https://github.com/user-attachments/assets/b6bb7ca3-ce78-4906-a6f8-cb1c9f549357" />

#### nft binary

<img width="1487" height="675" alt="network-create-execnft" src="https://github.com/user-attachments/assets/17cf5ce0-71e2-49ed-a546-ea8cfc479929" />

#### libnftables

<img width="1839" height="751" alt="network-create-cgonft" src="https://github.com/user-attachments/assets/c116f952-2a09-4469-a332-0b516687e12c" />

### Container start

#### iptables

<img width="1484" height="926" alt="container-start-iptables" src="https://github.com/user-attachments/assets/bd65fc07-2f09-4661-a8c3-c5b1862eb837" />

#### nft binary

<img width="1484" height="929" alt="container-start-execnft" src="https://github.com/user-attachments/assets/7db9cc78-2ab9-4a43-af51-b648b73f0bd3" />

#### libnftables

<img width="1843" height="1092" alt="Screenshot 2025-09-24 at 18 15 54" src="https://github.com/user-attachments/assets/c885ee07-3ab1-4de0-b3d6-311737c76fa2" />

</details>

**- How I did it**

Use cgo to call libnftables functions.

In packaging scripts/config, we'll need to add a build dependency on `libnftables-dev` and a installation dependency on `libnftables`.

**- How to verify it**

Unit tests use the library, integration tests (against the statically linked binary) use the nft tool.

**- Human readable description for the release notes**
```markdown changelog
- When dynamically linked, the Docker daemon now depends on libnftables.
```

robmry committed:
Use libnftables in dynamically linked binary

Signed-off-by: Rob Murray <rob.murray@docker.com>

corhere commented on daemon/libnetwork/internal/nftables/nftables_linux.go:
```diff
@@ -193,6 +191,7 @@ type table struct {
 	MustFlush      bool
 
 	applyLock sync.Mutex
+	nftHandle any // applyLock must be held to access
```
I have an idea on how to make this field concretely typed.
```suggestion
	nftHandle nftHandle // applyLock must be held to access
```

```go
// nft_cgo_linux.go
type nftHandle = *C.struct_nft_ctx
```
```go
// nft_exec_linux.go
type nftHandle = struct{}
```

robmry commented on daemon/libnetwork/drivers/bridge/internal/nftabler/nftabler.go:
```diff
@@ -72,6 +74,14 @@ func NewNftabler(ctx context.Context, config firewaller.Config) (*Nftabler, erro
+func (nft *Nftabler) Close() error {
+	_ = nft.table4.Close()
+	_ = nft.table6.Close()
+	return nil
```
Done

akerouanton commented on daemon/libnetwork/internal/nftables/nft_cgo_linux.go:
```diff
@@ -0,0 +1,78 @@
+	defer span.End()
+
+	if t.nftHandle == nil {
+		handle, err := newNftHandle()
```
How costly is it to instantiate a new `struct nft_ctx` and allocate new out/err buffers? I'm wondering if we could avoid `table.applyLock` if each call to `nftApply` is using its own `struct nft_ctx` as this lock serializes operations on independent networks / sandboxes.

robmry replied:
The `applyLock` isn't new ... it's there to protect the `table` while updates from a `Modifier` are applied to it, while an nftables command buffer is generated from the `Table`, and during `nftApply` (in case the update fails and changes to the `table` need to be rolled back).

The Daemon has a single table in the host's netns (plus tables in each container netns for DNS) with rules for all networks/endpoints. The table in the host netns has nftables base chains that do a single vmap lookup to decide whether the packet needs further processing. If they do, packets are processed by short chains dealing with a specific network. The alternative would be a table per-network, then it'd be possible to make updates for different networks in parallel. But then, instead of a single vmap lookup, each packet (including non-Docker packets) would need to be matched against base chain rules in each of the per-network tables.

akerouanton reviewed (approved):
LGTM, but wondering if we could do without `table.applyLock`.

robmry commented:
I'm still looking at the packaging changes needed for this ... please don't merge it yet.

robmry committed:
Add build tag "no_libnftables"

With this tag, a dynamically linked binary will exec
the nft tool instead of using cgo to call libnftables
directly.

Signed-off-by: Rob Murray <rob.murray@docker.com>
